### DISCUSSION TITLE
How to sort Planetoid dataset based on the degree values of nodes?

### DISCUSSION THREAD

**Discussion by OliverSummers** [pyg-team/pytorch_geometric#2952] [Q&A] [answered]
The Cora dataset has the following attributes:

`Data(edge_index=[2, 10556], test_mask=[2708], train_mask=[2708], val_mask=[2708], x=[2708, 1433], y=[2708])`

What I want to do is to sort the dataset based on the degree value of nodes. I code as follows:

    
    def sort_dataset(data):
       
        n = data.num_nodes
       
        degrees = degree(data.edge_index[1], n, dtype=torch.long)
        sorted, indices = torch.sort(degrees)
        new_x = data.x.detach().clone()
        new_y = data.y.detach().clone()

        for i in range(len(new_x)):
            new_x[i] = data.x[indices[i]]
            new_y[i] = data.y[indices[i]]

        data.x = new_x
        data.y = new_y

        return data

    dataset.transform = sort_dataset(data)

One thing missing in my code is rearranging data.edge_index as well. So based on the body of "sort_dataset" I shared above, how do I rearrange data.edge_index?

Thank you.

**Comment by rusty1s** [comment, marked as answer]
```python
deg = degree(edge_index[1], num_nodes, dtype=torch.long)
perm = deg.argsort()
x = x[perm]
edge_index = perm[edge_index]
```
should give you the desired result.

**Reply by OliverSummers** [reply]
Thank you. To get the new x and y, your code suggestion is much shorter than mine but I think there is still an issue with the new edge_index. When I sorted the dataset based on your suggestion and fed it into my GCN model, the accuracy was degraded to 23% from 81%. What is the issue in your opinion?

**Reply by rusty1s** [reply]
You are right. The above code has some bugs. This should fix it:
```python
inv_perm = torch.empty_like(perm)
inv_perm[perm] = torch.arange(perm.size(0))

x = x[perm]
edge_index = inv_perm[edge_index]
```

**Reply by OliverSummers** [reply]
Thank you very much. Your solutions are amazing. One more question, if I want to sort the dataset based on the page ranking of the nodes or any other requirement, how should we do it? Here we use `degree(edge_index[1], num_nodes, dtype=torch.long)` as our requirement is the degree value of nodes, but in other requirement cases, I don't know what to use instead of PyTorch `degree` function.

**Reply by rusty1s** [reply]
If you have the PageRank values of every node in a vector, then sorting based on these values is similar to sorting based on `deg` (via `argsort()`). We currently do not provide functionality to compute PageRank values though. As an alternative, you can convert your data to `networkx` and compute PageRank values from there.

**Reply by OliverSummers** [reply]
Thank you.
